src/lib/support/BufferReader.cpp

```diff
@@ -17,40 +17,46 @@
 
 #include "BufferReader.h"
 
+#include <lib/core/CHIPEncoding.h>
+
 #include <string.h>
+#include <type_traits>
 
 namespace chip {
 namespace Encoding {
 namespace LittleEndian {
 
 namespace {
-// These helper methods return void and put the value being read into an
+
+// This helper methods return void and put the value being read into an
 // outparam because that allows us to easily overload on the type of the
 // thing being read.
-void ReadHelper(const uint8_t *& p, uint8_t * dest)
-{
-    *dest = Read8(p);
-}
-void ReadHelper(const uint8_t *& p, uint16_t * dest)
-{
-    *dest = Read16(p);
-}
-void ReadHelper(const uint8_t *& p, uint32_t * dest)
+void ReadHelper(const uint8_t * p, bool * dest)
 {
-    *dest = Read32(p);
+    *dest = (*p != 0);
 }
-void ReadHelper(const uint8_t *& p, uint64_t * dest)
+
+template <typename T>
+void ReadHelper(const uint8_t * p, T * dest)
 {
-    *dest = Read64(p);
+    std::make_unsigned_t<T> result;
+    memcpy(&result, p, sizeof(result));
+    result = chip::Encoding::LittleEndian::HostSwap(result);
+
+    *dest = static_cast<T>(result);
```

Comment:
Casting to signed of things that are "negative" (i.e. out of range) has implementation-defined behavior until c++20...  

This should probably use `CastToSigned` for the casting.